Mathlib/Tactic/Linarith/SimplexAlgo/Gauss.lean

```diff
@@ -0,0 +1,135 @@
+/-
+Copyright (c) 2024 Vasily Nesterov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Vasily Nesterov
+-/
+import Mathlib.Tactic.Linarith.SimplexAlgo.Datatypes
+
+/-!
+# Gaussian Elimination algorithm
+
+The first step of `Linarith.SimplexAlgo.findPositiveVector` is finding initial feasible solution
+that is done by standard Gaussian Elimination algorithm implemented in this file.
+-/
+
+namespace Linarith.SimplexAlgo.Gauss
+
+/-- The mutable state for the `GaussM` monad. -/
+structure GaussState (n m : Nat) where
+  /-- Current matrix. -/
+  mat : Matrix n m
+  /-- Positions of units corresponding to basic variables in the `mat`. Used to obtain `Table` from
+  the processed matrix. -/
+  basicPositions : Array (Nat × Nat)
+  /-- Current row. -/
+  currentRow : Nat
+  /-- Current column. -/
+  currentColumn : Nat
+
+/-- The monad for the Gaussian Elimination algorithm. -/
+abbrev GaussM (n m : Nat) := StateM <| GaussState n m
+
+/-- Gets current row. -/
+def curRow {n m : Nat} : GaussM n m Nat := do
+  return (← get).currentRow
+
+/-- Gets current column. -/
+def curCol {n m : Nat} : GaussM n m Nat := do
+  return (← get).currentColumn
+
+/-- Increments current row. -/
+def incRow {n m : Nat} : GaussM n m Unit := do
+  set {← get with currentRow := (← curRow) + 1}
+
+/-- Increments current column. -/
+def incCol {n m : Nat} : GaussM n m Unit := do
+  set {← get with currentColumn := (← get).currentColumn + 1}
+
+/-- Pushes position to `basicPositions`. -/
+def pushToBasicPos {n m : Nat} (i j : Nat) : GaussM n m Unit := do
+  set {← get with basicPositions := (← get).basicPositions.push ⟨i, j⟩}
+
+/-- Finds the first row starting from the current column with nonzero element in current column. -/
+def findNonzeroRow {n m : Nat} : GaussM n m <| Option Nat := do
+  for i in [← curRow:n] do
+    if (← get).mat[i]![← curCol]! != 0 then
+      return i
+  return .none
+
+/-- Swaps two rows. -/
+def swapRows {n m : Nat} (i j : Nat) : GaussM n m Unit := do
+  if i == j then
+    return
+  let swapped : Matrix n m := ⟨(← get).mat.data.swap! i j⟩
+  set {← get with mat := swapped}
+
+/-- Subtracts `i`-th row * `coef` from `j`-th row. -/
+def subtractRow {n m : Nat} (i j : Nat) (coef : Rat) : GaussM n m Unit := do
+  let new_row := (← get).mat[j]!.zip (← get).mat[i]! |>.map fun ⟨x, y⟩ => x - coef * y
+  let subtractedMat : Matrix n m := ⟨(← get).mat.data.set! j new_row⟩
+  set {← get with mat := subtractedMat}
```

Comment:
nit: I think using `modify` here is slightly more idiomatic; perhaps as `modify ({ \. with at := subtractedMat })